### DISCUSSION TITLE
Type prisma client dynamically when using multiple schema

### DISCUSSION THREAD

**Discussion by KGALLET** [prisma/prisma#17270] [Q&A]
Hello,

First of all, I think that's more a Typescript issue than a Prisma one

I'm working on a psql database using multischema: 
![image](https://user-images.githubusercontent.com/14168629/212089912-7e43859d-cfb5-4b80-acdb-abd6f3409f77.png)

All the models that are schema specific are the same (i.e. have the same column, and type), we use the multi schema to ensure we fetch / save the data in the proper schema. 

I want to instantiate my prisma client dynamically, using a key that is linked to my schema. 
![image](https://user-images.githubusercontent.com/14168629/212090696-c49e3885-cf50-485c-beb8-4df9d1c80684.png)

I did not find a way to type the client properly 
![image](https://user-images.githubusercontent.com/14168629/212091007-8cbb7fba-b515-44a2-b1bc-c452910b8757.png)

Here is the error I get : 
![image](https://user-images.githubusercontent.com/14168629/212091840-6111387d-0647-4ae4-8848-c4c8ae08cc8d.png)

Obvisouly if i specify the client I use for the `findUnique` method, typescript infer the type and I don't have any problem:
![image](https://user-images.githubusercontent.com/14168629/212091186-36ece980-185a-4962-a1cb-e28edda587c6.png)

I can also create a repository that extends an abstract class for each schema, and instantiate the proper client in the constructor, but the code maintenance / duplication of code would be too high for my use case. 

Do you know how can I avoid to type as any my client ?
![image](https://user-images.githubusercontent.com/14168629/212091683-b750a2ad-6f36-4baf-b004-c9924973f638.png)

Thank's

**Comment by pantharshit00** [comment]
Hello @KGALLET.


Can you please provide your complete schema here? Also, please paste code snippets present in the screenshots so that we can quickly try to reproduce this instead of trying to copy code from images.

**Reply by KGALLET** [reply]
Of course,

Here you are : 

`schema.prisma` 

```
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["multiSchema", "extendedWhereUnique"]
}

datasource db {
  provider          = "postgresql"
  url               = env("DATABASE_URL")
  shadowDatabaseUrl = env("SHADOW_DATABASE_URL")
  schemas           = ["inter", "allemagne", "autriche"]
}

model AustriaCustomer {
  customerId              String           @id @default(uuid()) @map("customer_id") @db.VarChar(36)
  firstname               String?          @map("firstname") @db.VarChar(36)
  lastname                String           @map("lastname") @db.VarChar(36)
  username             String.            @map("username") @db.VarChar(36)

  @@map("customer")
  @@schema("autriche")
}

model AustriaAddress {
  addressId  String @id @default(uuid()) @map("address_id") @db.VarChar(36)
  customerId String @map("customer_id") @db.VarChar(36)

  type        String  @db.VarChar(15)
  address    String? @db.VarChar(50)
  postcode    String  @db.VarChar(15)
  city        String  @db.VarChar(50)
  countryCode String? @map("country_code") @db.VarChar(2)

  updatedAt DateTime @default(now()) @map("updated_at") @db.Timestamp()
  validated Boolean  @default(true) @db.Boolean

  GermanyCustomer GermanyCustomer @relation("address", fields: [customerId], references: [customerId])

  @@map("address")
  @@schema("autriche")
}


model GermanyCustomer {
  customerId              String           @id @default(uuid()) @map("customer_id") @db.VarChar(36)
  firstname               String?          @map("firstname") @db.VarChar(36)
  lastname                String           @map("lastname") @db.VarChar(36)

  @@map("customer")
  @@schema("allemagne")
}

model GermanyAddress {
  addressId  String @id @default(uuid()) @map("address_id") @db.VarChar(36)
  customerId String @map("customer_id") @db.VarChar(36)

  type        String  @db.VarChar(15)
  address    String? @db.VarChar(50)
  postcode    String  @db.VarChar(15)
  city        String  @db.VarChar(50)
  countryCode String? @map("country_code") @db.VarChar(2)

  updatedAt DateTime @default(now()) @map("updated_at") @db.Timestamp()
  validated Boolean  @default(true) @db.Boolean

  GermanyCustomer GermanyCustomer @relation("address", fields: [customerId], references: [customerId])

  @@map("address")
  @@schema("allemagne")
}
```

`client.ts`

``` 
export type PrismaClientKey = 'AT' | 'BE' | 'DE' | 'ES' | 'FR' | 'IT' | 'PT'

const customerClients = {
  AT: prismaClient.austriaCustomer,
  BE: prismaClient.belgiumCustomer,
  DE: prismaClient.germanyCustomer,
  ES: prismaClient.spainCustomer,
  FR: prismaClient.franceCustomer,
  IT: prismaClient.italyCustomer,
  PT: prismaClient.portugalCustomer,
}

export const CustomerClient = { ...customerClients }
```

`repository.ts`

```
const findCustomerById = async (customerId: string, countryCode: PrismaClientKey): Promise<Customer> => {
  Logger.logInfo({ message: `Get customer by id ${customerId} in country: ${countryCode}` })
  const customerEntity = await (CustomerClient[countryCode] **as any**).findUnique({
    include: {
      address: true,
      email: true,
      phone: true,
    },
    where: { customerId },
  })
```

**Comment by ludralph** [comment]
Hi @KGALLET 👋 
You can follow these steps to type your Prisma client dynamically

1. Define a type alias that captures the common structure of client properties for different schemas.
```ts
export type CustomerClientMap = {
  AT: Prisma.AustriaCustomerClient;
  BE: Prisma.BelgiumCustomerClient;
  // ...other countries
};
```

2. Create a helper function to assert the type of the client based on the country code.
```ts
const getClient = <T extends keyof CustomerClientMap>(countryCode: T): Prisma[CustomerClientMap[T]] => {
  return CustomerClient[countryCode] as Prisma[CustomerClientMap[T]];
};
```

3. Access and use the client with proper typing in your repository.
```ts
const findCustomerById = async (customerId: string, countryCode: keyof CustomerClientMap): Promise<Customer> => {
  Logger.logInfo({ message: `Get customer by id ${customerId} in country: ${countryCode}` });
  const client = getClient<keyof CustomerClientMap>(countryCode);
  const customerEntity = await client.findUnique({
    include: {
      address: true,
      email: true,
      phone: true,
    },
    where: { customerId },
  });
};
```

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
